### PR TITLE
Shooting zones

### PR DESCRIPTION
**Pull request by YoniKiriaty**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/subsystems/shooter/pivot/Pivot.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import edu.wpi.first.math.geometry.Rotation2d;
 import edu.wpi.first.math.geometry.Rotation3d;
 import org.littletonrobotics.junction.Logger;
+import edu.greenblitz.robotName.subsystems.shooter.pivot.PivotConstants;
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/shooter/pivot/Pivot.java:10]: Why?

**Review by yonnbro — COMMENTED** [review]
I need an explanation about what is shooting zone and the wrapper zone

**Comment by YoniKiriaty** [comment]
> I need an explanation about what is shooting zone and the wrapper zone

The problem was that if the robot was too close to it's shooting zone position it would not move to it because of pathplanner 🙄.
The way it is fixed is by making a wrapper zone around a small zone, which if the robot is inside it it would behave how it should be when it is used inside a zone, but if it is outside the wrapper zone it tries to go the the bound of the optimal zone, which is included inside the wrapper zone. Also the shooting zone was made so it is not a full circle but a specific portion determined by an angle:
quarter circle

The close shooting zone is for the possibility that we can't shoot too far/consistent enough.

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java [hunk 2]
```diff
@@ -10,17 +10,59 @@ public class ShootingPositionConstants {
...
-    public static final Rotation2d LOWER_ANGLE_LIMIT = Rotation2d.fromDegrees(-50);
+    public static final Rotation2d LEGAL_UPPER_ANGLE_LIMIT = Rotation2d.fromDegrees(30);
 
-    public static final Rotation2d UPPER_ANGLE_LIMIT = Rotation2d.fromDegrees(30);
+    public static final Rotation2d OPTIMAL_LOWER_ANGLE_LIMIT = Rotation2d.fromDegrees(-30);
 
-    public static ShootingZone SHOOTING_ZONE = new ShootingZone(
+    public static final Rotation2d OPTIMAL_UPPER_ANGLE_LIMIT = Rotation2d.fromDegrees(10);
+
+    public static final double LEGAL_SHOOTING_ZONE_RADIUS_IN_METERS = 3;
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:21]: Make the legal radius const “bound” to the optimal radius const

Comment (yonnbro) [reply src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:21]: So just by changing the optimal radius the legal radius updates as well

Comment (yonnbro) [reply src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:21]: (Doesnt have to be at runtime, consts are good

**Comment by YoniKiriaty** [comment]
I went for a different implementation that doesnt use the wrapper because its simpler and has no issues.
now if its too close it goes to the middle of the shooting zone

**Comment by yonnbro** [comment]
> I realized i could go for a different implementation that doesnt use the wrapper because its simpler and has no issues.
> now if its too close it goes to the middle of the shooting zone

Please talk to us before changing things we agreed on… I really really don’t like this behavior… it will create unexpected long chassis for the driver, the previous solution is better IMO

**Comment by YoniKiriaty** [comment]
> > I realized i could go for a different implementation that doesnt use the wrapper because its simpler and has no issues.
> > now if its too close it goes to the middle of the shooting zone
> 
> Please talk to us before changing things we agreed on… I really really don’t like this behavior… it will create unexpected long chassis movements for the driver, the previous solution is better IMO

ok

**Review by yonnbro — APPROVED** [review]

File: src/main/java/edu/greenblitz/robotName/commands/shooter/ShootFromCloseRange.java [hunk 3]
```diff
@@ -1,20 +1,13 @@
...
-import edu.greenblitz.robotName.commands.swerve.RotateToAngle;
-import edu.greenblitz.robotName.subsystems.shooter.pivot.PivotConstants;
-import edu.greenblitz.robotName.subsystems.swerve.chassis.ChassisConstants;
-import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
-import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+import edu.greenblitz.robotName.commands.shooter.shootingState.GoToShootingStateAndShoot;
+import edu.greenblitz.robotName.shootingStateService.ShootingPositionConstants;
 
-public class ShootFromCloseRange extends SequentialCommandGroup {
+public class ShootFromCloseRange extends GoToShootingStateAndShoot {
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/commands/shooter/ShootFromCloseRange.java:6]: i think that it is not needed

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java [hunk 4]
```diff
@@ -10,17 +10,56 @@ public class ShootingPositionConstants {
 
     public static final Translation2d CENTER_OF_SHOOTING_ZONE = FieldConstants.MIDDLE_OF_SPEAKER_POSITION.toTranslation2d();
 
-    public static final double SHOOTING_ZONE_RADIUS_IN_METERS = 4;
+    public static final Rotation2d LEGAL_LOWER_ANGLE_LIMIT = Rotation2d.fromDegrees(-50);
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:13]: why in negative numbers, please convert to positive

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java [hunk 5]
```diff
@@ -10,17 +10,56 @@ public class ShootingPositionConstants {
...
+                CENTER_OF_SHOOTING_ZONE,
+                LEGAL_SHOOTING_ZONE_RADIUS_IN_METERS,
+                FieldConstants.FieldZones.RESTRICTED_BOUNDS,
+                LEGAL_LOWER_ANGLE_LIMIT,
+                LEGAL_UPPER_ANGLE_LIMIT,
+                LEGAL_SHOOTING_ZONE
+        );
+    }
+
+    public static final double OPTIMAL_SHOOTING_ZONE_RADIUS_IN_METERS = 2.5;
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingPositionConstants.java:36]: spaces

File: src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingStateCalculations.java [hunk 6]
```diff
@@ -2,40 +2,41 @@
...
 
     private static Pose2d getRobotPose() {
         return SwerveChassis.getInstance().getRobotPose();
     }
 
-    public static Translation2d getRobotTargetTranslation() {
-        return SHOOTING_ZONE.getClosestCirclePosition(getRobotPose().getTranslation());
+    public static Translation2d getRobotTargetTranslation(ShootingZone zone) {
+        Translation2d closestPosition = zone.getClosestCirclePosition(getRobotPose().getTranslation());
+        if (closestPosition.getDistance(getRobotPose().getTranslation()) > SwerveChassis.TRANSLATION_TOLERANCE &&
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/shootingStateService/ShootingStateCalculations.java:27]: add {

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Pull request merged**